### ISSUE TITLE
latency() got an unexpected keyword argument 'num_outputs' and latency() got multiple values for argument 'num_steps'

### ISSUE THREAD

**Issue by Dola47** [jeshraghian/snntorch#48]
* snntorch version: 0.2.11
* Python version:  3.7.10

### Description

In the latest updates, I see that the `num_output` has been removed from the `spikegen`. No clue why!

Moreover, also when I just decide about removing the `num_output` parameter. I get another error from the `num_steps` when I set it to any int value?

BTW, the same exists if I tried to do rate encoding instead of latency encoding.

### What I Did

[Here:](https://colab.research.google.com/drive/1XPWRmcuHAfTFFqw3UTKcZiQe249x0qcz?usp=sharing) you will find a quick ipynb file that shows the errors.

**Comment by jeshraghian** [comment]
> In the latest updates, I see that the num_output has been removed from the spikegen

It's been replaced with `num_classes` instead - just making it match PyTorch a little better.

The changes have been documented [here](https://snntorch.readthedocs.io/en/latest/snntorch.spikegen.html), or you can take a look at several examples by using `help(spikegen.rate)` etc. 

I have requested access to the ipynb file as I cannot tell what the error from `num_steps` is.

**Comment by Dola47** [comment]
@jeshraghian 

Can you check the ipynb now [Here](https://colab.research.google.com/drive/1XPWRmcuHAfTFFqw3UTKcZiQe249x0qcz?usp=sharing)?

**Comment by jeshraghian** [comment]
Gotcha! 

This should work. 

Note that `spikegen.latency` is now intended for input features. So including the number of outputs/classes would be meaningless.

You can instead separately enter that into `spikegen.targets_latency`. 



```
# Spiking Data
spike_data = spikegen.latency(data_it, num_steps=num_steps, tau=5, threshold=0.01, linear=True, normalize=True, clip=True)

spike_targets = spikegen.targets_latency(targets_it, num_classes=num_outputs, num_steps=num_steps, tau=5, threshold=0.01,
                                                     linear=True, normalize=True, clip=True)
```

**Comment by Dola47** [comment]
Hi @jeshraghian,

I just noticed one thing, if we coded the targets as you have done here, the output will be of the shape of [num_steps, batch_size, number_of_classes]. Even if you iterated over the num of steps as it is done here: 

```
        for step in range(num_steps):
            loss_val += loss_fn(log_p_y[step], spike_targets[step])
```
We will have 2d targets and we will not be able to use it with the NNLLoss for example, or even the MSELoss. So I wonder have you expect the spiking targets to be used within your code.

Moreover, for the accuracy calculation if we are counting the number of spikes because you are summing over the temporal domain as it is done here: 

```
def print_batch_accuracy(data, targets, train=False):
    output, mem_rec = net(data.view(num_steps, batch_size, 1, 28, 28))
    _, idx = output.sum(dim=0).max(1)
```
Then we do not need the spiking targets anymore when we call `print_batch_accracy` but we also need the ground truth targets of the batch without any encoding?

Let me know your thoughts!

Thanks.

**Comment by jeshraghian** [comment]
Hi @Dola47,

Just a reminder that issues are intended for reporting bugs. When in doubt, open a **new** thread in **discussions** so that users with the same question can refer to that easily. If it turns out it's a bug, then we can convert the discussion into an issue! 

> We will have 2d targets and we will not be able to use it with the NNLLoss for example, or even the MSELoss. So I wonder how you expect the spiking targets to be used within your code?

A 2d target of [batch_size x num_outputs] can be used to apply a separate MSELoss to each output neuron.
Such a target represents the 'goal' of the membrane potential of each neuron.
The 10 losses are then summed together both over time, and over neurons. 
You can can check this notebook out as an example:

https://colab.research.google.com/drive/1j9-ddyadCMyn6N12zjPmu98VAmYF3YIl?usp=sharing

Another note; I've since added a couple other neuron types.

In my own experiments, `Leaky` consistently outperforms `Synaptic` (used to be `Stein`) for the right set of parameters.
The only difference is that Leaky takes an input, integrates it into membrane potential (decay at rate beta), and then emits a spike.
Stein/Synaptic take an input, integrate it into a synaptic current (decay at rate alpha), which is integrated again into membrane potential (decay at rate beta), and then emits a spike. 
The notebook above shows a code sample with `Leaky`. 

Next question -

> Then we do not need the spiking targets anymore when we call print_batch_accracy but we also need the ground truth targets of the batch without any encoding?

Yeah - you are absolutely right. No need for the converted targets for this particular task. You only compare the raw ground truth targets to the spike count. 

There are obviously other tasks that might require the converted targets to measure accuracy, e.g., if you are trying to encourage a neuron to fire at very specific times, or the goal is to achieve a given pattern/evolution of membrane potential.
